master/static/migrations/20230718150024_trial-one-to-many-tasks.tx.up.sql

```diff
@@ -0,0 +1,76 @@
+CREATE TABLE public.trial_id_task_id (
+    trial_id integer REFERENCES trials(id) ON DELETE CASCADE NOT NULL,
+    task_id text REFERENCES tasks(task_id) ON DELETE CASCADE NOT NULL UNIQUE,
+    task_run_id integer DEFAULT 0 NOT NULL, -- Starts at 0. Trial gets restarted goes to 1.
+    PRIMARY KEY(trial_id, task_id)
+);
+
+CREATE INDEX idx_trial_id_task_id ON public.trial_id_task_id USING btree (trial_id);
```

Comment:
we should already get a (trial_id, task_id) index from the primary key, so we may not need this. i might check the plans.

Reply:
Removed the ```trial_id, task_id``` index.

Is it correct that there is still a benefit to having single column indexes despite being part apart of the composite primary key index like 

```
CREATE INDEX idx_trial_id_task_id ON public.trial_id_task_id USING btree (trial_id);
```

```
CREATE INDEX idx_trial_id_task_id_task_id ON public.trial_id_task_id USING btree (task_id);
```

Reply:
I would guess that any query on that table is going for the mapping, so an index on `(task_id, trial_id)` and `(trial_id, task_id)` and nothing more will be optimal because then I think every query can do `Index Only Scan` and never need to go to actual table.

Reply:
As opposed to indexes on the single columns where they get pages or ctids for tuples and have to go to the table to get the values for task_id/trial_id for the task_id/trial_id they want.

Reply:
https://postgres.ai/blog/20220106-explain-analyze-needs-buffers-to-improve-the-postgres-query-optimization-process / https://www.cybertec-postgresql.com/en/how-to-interpret-postgresql-explain-analyze-output/ are both good articles, btw

Reply:
Changed to adding a task_id/trial_id with the primary key index also for trial_id/task_id

Interesting buffers should be a metric to go by rather than timing. I wonder how much CLUSTER would help us in some places.

I do find myself using https://explain.depesz.com/ a lot.